platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingArtifactVariantSelector.java

```diff
@@ -80,78 +80,27 @@ private ResolvedArtifactSet doSelect(
         boolean allowNoMatchingVariants
     ) {
         AttributeMatcher matcher = attributeSchemaServices.getMatcher(consumerSchema, producer.getProducerSchema());
-        ImmutableAttributes componentRequested = attributesFactory.concat(requestAttributes, producer.getOverriddenAttributes());
-        final List<ResolvedVariant> variants = producer.getCandidates();
-
-        List<? extends ResolvedVariant> matches = matcher.matchMultipleCandidates(variants, componentRequested);
-        if (matches.size() == 1) {
-            return matches.get(0).getArtifacts();
-        } else if (matches.size() > 1) {
-            throw failureProcessor.ambiguousArtifactsFailure(matcher, producer, componentRequested, matches);
-        }
-
-        // We found no matches. Attempt to construct artifact transform chains which produce matching variants.
-        List<TransformedVariant> transformedVariants = consumerProvidedVariantFinder.findTransformedVariants(variants, componentRequested);
-
-        // If there are multiple potential artifact transform variants, perform attribute matching to attempt to find the best.
-        if (transformedVariants.size() > 1) {
-            transformedVariants = tryDisambiguate(matcher, transformedVariants, componentRequested);
+        ImmutableAttributes targetAttributes = attributesFactory.concat(requestAttributes, producer.getOverriddenAttributes());
+
+        // Check for matching variant without using artifact transforms.  If we found only one, return it.  If we found multiple matches, that's ambiguity.
+        List<ResolvedVariant> matchingVariants = matcher.matchMultipleCandidates(producer.getCandidates(), targetAttributes);
+        if (matchingVariants.size() == 1) {
+            return Iterables.getOnlyElement(matchingVariants).getArtifacts();
+        } else if (matchingVariants.size() > 1) {
+            throw failureHandler.ambiguousArtifactsFailure(matcher, producer, targetAttributes, matchingVariants);
         }
 
-        if (transformedVariants.size() == 1) {
-            TransformedVariant result = transformedVariants.get(0);
-            return producer.transformCandidate(result.getRoot(), result.getTransformedVariantDefinition());
-        }
-
-        if (!transformedVariants.isEmpty()) {
-            throw failureProcessor.ambiguousArtifactTransformsFailure(producer, componentRequested, transformedVariants);
+        // We found no matching variant.  Attempt to select a chain of transformations that produces a suitable virtual variant.
+        Optional<TransformedVariant> selectedTransformationChain = transformationChainSelector.selectTransformationChain(producer, targetAttributes, matcher);
```

Comment:
Using `Optional` here with one `isPresent()` and two calls to `get()` seems noisy here compared to a simple null check

Reply:
Left for a larger discussion about the proper use of `null` vs. `Optional`.